File: hanoi/backtracking/backtracking.py
from hanoi.utils.hanoi import *
import time


def start_backtracking(pegs, pieces):
    def go(state, current_piece_index, previous_states):
        if is_final_state(state):
            print("Found state! " + str(state))

        for peg in range(1, state[0] + 1):
            if is_valid(state, current_piece_index, peg):
                new_state = transition(state, current_piece_index, peg)
                print("Current state " + str(new_state))
                print("Current piece index " + str(current_piece_index))
                print("Previous states " + str(previous_states))
                time.sleep(2)
                for x in range(1, new_state[0] + 1):
                    if x != current_piece_index:
                        return go(new_state, x, previous_states +  [new_state])

    initial_state = initialize(pegs, pieces)
    return go(initial_state, 1, [])


print(start_backtracking(3, 3))

File: hanoi-towers/src/backtracking/Backtracking.scala
package backtracking

import utils.Utils._

import scala.annotation.tailrec

object Backtracking {
  type States = List[List[Int]]

  def solveHanoi(pegs: Int, pieces: Int): List[States] = {
    def go(currentState: List[Int], currentPieceIndex: Int, previousStates: States): List[States] = {
      if (isFinalState(currentState)){
        print("FOUND")
        List(previousStates)
      }
      else {
        //println("Current state " + currentState)

        //all possible routes that could be taken from the current piece index in current state
        val validStates = (1 to currentState.head).toList.filter(e => isValid(currentState, currentPieceIndex, e))

        println("Valid states " + validStates)
        //a list of all possible states, not including previous states
        val possibleStates = validStates
          .map(p => transition(currentState, currentPieceIndex, p))
          .filterNot(previousStates.contains(_))

        println("Possible states " + possibleStates)

        println("Previous states " + previousStates)

        //Thread.sleep(3000)
        possibleStates.flatMap(s => (1 to currentState.head)
          .toList
          .filterNot(_ == currentPieceIndex)
          .flatMap(piece => go(s, piece, previousStates :+ s)))
      }
    }

    go(initialize(pegs, pieces), 1, List(List.empty))
  }
}
